master/buildbot/process/build.py

```diff
@@ -291,8 +283,59 @@ def startBuild(self, build_status, workerforbuilder):
             log.err(Failure())
             self.buildFinished(['Build.setupBuild', 'failed'], EXCEPTION)
             return
+
         # flush properties in the beginning of the build
         yield self._flushProperties(None)
+
+        try:
+            ready = yield workerforbuilder.prepare(self)
+        except Exception:
+            log.err(Failure(), 'while preparing workerforbuilder:')
+            self.buildFinished(["worker", "not", "available"], RETRY)
+            return
+
+        # If prepare returns True then it is ready and we start a build
+        # If it returns false then we don't start a new build.
+        if not ready:
+            log.msg("worker %s can't build %s after all; retrying the "
+                    "build" % (self, workerforbuilder))
+            self.stopped = True
+            self.buildFinished(["worker", "not", "available"], RETRY)
```

Comment:
No. The only thing waiting for `buildStarted` is a last-ditch exception logger, and `buildFinished` already has one of those. In addition, `buildFinished` is designed to be called in response to events (where there is nothing waiting on the result), instead of as part of a sequence of action that need to be done serially. If I was writing `buildFinshed` without `inlineCallbacks` (i.e. with raw deferreds), I wouldn't have it return a deferred, so the fact that it does is merely an implementation detail.